### ISSUE TITLE
Shallow Water demo, and water normal methods

### ISSUE THREAD

**Issue by jasonsturges** [openfl/away3d-samples#3]
Is the shallow water demo functional?  HTML5, Mac, and Flash targets aren't rendering the water, and mouse interaction is freezing the water instead of rippling:

![shallow-water](https://cloud.githubusercontent.com/assets/1213591/24391298/c9a9c4dc-1353-11e7-89b4-2ee0a388e1a9.gif)

In general, water normals seem to have issue on c++ targets.

In my [OpenFL Away3D Terrain Demo](https://github.com/jasonsturges/openfl-away3d-terrain-demo), water is spectacular on HTML5 and Flash targets, but fails to render on c++.

![water-normals](https://cloud.githubusercontent.com/assets/1213591/24392050/dd80538c-1357-11e7-85e7-af888abe3880.gif)

Water implemented as:

```
    waterMethod = new SimpleWaterNormalMethod(
        Cast.bitmapTexture("assets/water/water_normals.jpg"),
        Cast.bitmapTexture("assets/water/water_normals.jpg"));

    waterMaterial = new TextureMaterial(new BitmapTexture(new BitmapData(512, 512, true, 0xaa404070)));
    waterMaterial.alphaBlending = true;
    waterMaterial.lightPicker = lightPicker;
    waterMaterial.repeat = true;
    waterMaterial.normalMethod = waterMethod;
    waterMaterial.addMethod(new EnvMapMethod(cubeTexture));
    waterMaterial.specularMethod = fresnelMethod;
    waterMaterial.gloss = 100;
    waterMaterial.specular = 1;
```

Are these known issues?

Probably unrelated, but the Terrain Demo c++ target is showing a corrupt buffer:

![terrain-demo](https://cloud.githubusercontent.com/assets/1213591/24391430/b0df0b00-1354-11e7-8a53-ad9593c94dfd.png)

Configuration:
away3d: 5.0.1
hxcpp: 3.4.64
lime: 4.0.2
openfl: 4.9.1

**Comment by Greg209** [comment]
Using the latest dev builds of lime and openfl as of today, I managed to get the Terrain demo up and running fine targetting C++ & neko (on OSX at least).
![image](https://cloud.githubusercontent.com/assets/550301/24423440/da5c5bb8-13f4-11e7-9fb8-f155d273df9f.png)

Could you check against the latest dev builds?

As for the Shallow water demo, looking at the demo code, it appears there is a dependency PixelBender shaders to manipulate the water ripples. These currently aren't implemented in OpenFL and the embeds are commented out so wouldn't even work on the flash target.

https://github.com/openfl/away3d-samples/blob/master/advanced/ShallowWaterDemo/Source/shallowwater/ShallowFluid.hx#L65-L75

So until PixelBender shaders are implemented or the shaders are re-worked in a more OpenFL way it's not going to work as in the original demo.

I'll leave this open for now so it's still on the radar.

**Comment by jgranick** [comment]
I think there's a regression in Lime 4.x that I'm trying to resolve, causing the corruption issue

**Comment by Greg209** [comment]
@jgranick One thing I've seen - probably not related though - is that performance (having tried a few of the away3d samples) of C++ and Neko on the mac is very bad in comparison to webgl - even non-debug versions. ShallowWater (on old 2009 MacBookPro) on Neko (non-debug) takes ages to start, may 20-30secs of black screen then 1-2FPS, C++ (debug version) gives 15FPS whilst HTML5 gives a pleasant 52FPS (debug as well).

I've yet to look into potential reasons why but struggled to create an xcode project to debug with. Need to look into using hxScout again as well - it's been a while.

Have you seen similar poor performance on C++ or Neko targets?

**Comment by jgranick** [comment]
The latest HXCPP and Lime work with HxScout, use `-Dtelemetry` when you build to enable it.

The Stage3D code has been undergoing accuracy focus, and likely has some low-hanging fruit for performance optimization. I've seen it vary on the system, it could also be an unoptimized code path hitting Cairo too often. Is the demo you're running using AwayFPS or some other object with shapes?

**Comment by Greg209** [comment]
Yes - using AwayFPS. I'll try and get some time to dig into it and see where the bottleneck(s) are.

**Comment by jgranick** [comment]
I found a regression in render-to-texture support in OpenFL, here's the TerrainDemo after the fix:

![image](https://cloud.githubusercontent.com/assets/833997/24433207/f8028ffc-13db-11e7-9994-b6bc979c6f1f.png)

**Comment by jgranick** [comment]
...and testing the C++ target on your sample as well:

![image](https://cloud.githubusercontent.com/assets/833997/24433397/24dfccaa-13dd-11e7-8a63-d00ff003e617.png)

**Comment by jasonsturges** [comment]
All around awesome - thanks to you both.

Per performance of c++ targets on mac, in my experience they load quick and are responsive achieving 60 fps.  Though I haven't pressed any limits, I haven't observed any substandard performance in either debug or release builds.

Not quite up to speed, but I can take a stab at profiling on Mac.

OS X El Capitan v10.11.6, retina, 2.4 GHz Intel Core i7, 8 GB 1600 MHz DDR3, NVIDIA GeForce GT 650M 1024 MB

**Comment by jgranick** [comment]
@Greg209 the performance might be slower HXCPP debug performance, I have optimizations in Lime and OpenFL that flag core classes for release even in debug builds, which should help (but only workshows in Haxe 3.4+)
